src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FallbackLinker.java

```diff
@@ -170,89 +166,105 @@ private static Object doDowncall(SegmentAllocator returnAllocator, Object[] args
         int numArgs = argLayouts.size();
         MemoryLayout retLayout = data.returnLayout();
         try (Arena upcallArena = Arena.openConfined()) {
             MemorySegment argsSeg = MemorySegment.ofAddress(argPtrs.address(), numArgs * ADDRESS.byteSize(), upcallArena.scope());
-            MemorySegment retSeg = data.rType() != TypeClass.VOID
+            MemorySegment retSeg = retLayout != null
                 ? MemorySegment.ofAddress(retPtr.address(), retLayout.byteSize(), upcallArena.scope())
                 : null;
 
             Object[] args = new Object[numArgs];
             for (int i = 0; i < numArgs; i++) {
                 MemoryLayout argLayout = argLayouts.get(i);
-                TypeClass argType = data.aTypes().get(i);
                 MemorySegment argPtr = MemorySegment.ofAddress(argsSeg.getAtIndex(JAVA_LONG, i), argLayout.byteSize(), upcallArena.scope());
 
-                args[i] = readValue(argPtr, argType, argLayout);
+                args[i] = readValue(argPtr, argLayout);
             }
 
             Object result = data.target().invokeWithArguments(args);
 
-            writeValue(data.rType(), result, data.returnLayout(), retSeg);
+            writeValue(result, data.returnLayout(), retSeg);
         }
     }
 
     // where
-    private static void writeValue(TypeClass type, Object arg, MemoryLayout layout, MemorySegment argSeg) {
-        writeValue(type, arg, layout, argSeg, addr -> {});
+    private static void writeValue(Object arg, MemoryLayout layout, MemorySegment argSeg) {
+        writeValue(arg, layout, argSeg, addr -> {});
     }
 
-    private static void writeValue(TypeClass type, Object arg, MemoryLayout layout, MemorySegment argSeg,
+    private static void writeValue(Object arg, MemoryLayout layout, MemorySegment argSeg,
                                    Consumer<MemorySegment> acquireCallback) {
-        switch (type) {
-            case BOOLEAN -> argSeg.set((ValueLayout.OfBoolean) layout, 0, (Boolean) arg);
-            case BYTE -> argSeg.set((ValueLayout.OfByte) layout, 0, (Byte) arg);
-            case CHAR -> argSeg.set((ValueLayout.OfChar) layout, 0, (Character) arg);
-            case SHORT -> argSeg.set((ValueLayout.OfShort) layout, 0, (Short) arg);
-            case INT -> argSeg.set((ValueLayout.OfInt) layout, 0, (Integer) arg);
-            case LONG -> argSeg.set((ValueLayout.OfLong) layout, 0, (Long) arg);
-            case FLOAT -> argSeg.set((ValueLayout.OfFloat) layout, 0, (Float) arg);
-            case DOUBLE -> argSeg.set((ValueLayout.OfDouble) layout, 0, (Double) arg);
-            case ADDRESS -> {
-                MemorySegment addrArg = (MemorySegment) arg;
-                acquireCallback.accept(addrArg);
-                argSeg.set((ValueLayout.OfAddress) layout, 0, addrArg);
-            }
-            case SEGMENT -> argSeg.copyFrom((MemorySegment) arg); // by-value struct
-            case VOID -> {}
+        if (layout instanceof ValueLayout.OfBoolean bl) {
```

Comment:
Do you think we lose a lot of performance if we don't switch? E.g. use the underlying var handle (which is stored in the value layout) and access it in an inexact mode?

Reply:
I considered that, but thought inexact access would be a lot worse in terms of performance since then we lose all inlining, and have to go through `asType` a lot of the time (since we probably won't see the same layout type all the time).

I think what I ended up doing is a nice balance between performance and simplicity, and looks kind of like a 'template' for using type switch in the future.